File: backend/app/routers/workflows.py
from fastapi import APIRouter, HTTPException, Depends, Body
from typing import List, Dict, Any, Optional
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.workflow import Workflow
from app.models.user import User
from app.services.workflow_service import WorkflowService
from app.services.subscription_service import SubscriptionService
from app.routers.users import get_current_user
from pydantic import BaseModel
from datetime import datetime

router = APIRouter()

# Modèles de données Pydantic pour la validation
class NodeModel(BaseModel):
    id: str
    type: str
    position: Dict[str, float]
    data: Dict[str, Any]
    width: int = None
    height: int = None
    
    class Config:
        arbitrary_types_allowed = True

class EdgeModel(BaseModel):
    id: str
    source: str
    target: str
    sourceHandle: str = None
    targetHandle: str = None
    type: str = None
    animated: bool = False
    label: str = None
    data: Dict[str, Any] = None
    
    class Config:
        arbitrary_types_allowed = True

class WorkflowCreateModel(BaseModel):
    name: str
    description: str = None
    nodes: List[Dict[str, Any]] = []
    edges: List[Dict[str, Any]] = []
    createdAt: str = None
    is_shared: bool = False
    is_template: bool = False
    
    class Config:
        arbitrary_types_allowed = True

class WorkflowUpdateModel(BaseModel):
    name: str = None
    description: str = None
    nodes: List[Dict[str, Any]] = None
    edges: List[Dict[str, Any]] = None
    is_shared: bool = None
    is_template: bool = None
    
    class Config:
        arbitrary_types_allowed = True

class WorkflowResponseModel(BaseModel):
    id: str
    name: str
    description: str = None
    owner_id: str
    company_id: str = None
    nodes: List[Dict[str, Any]]
    edges: List[Dict[str, Any]]
    createdAt: str = None
    is_shared: bool
    is_template: bool
    storage_size: int
    created_at: str
    updated_at: str
    
    class Config:
        arbitrary_types_allowed = True

# Endpoints
@router.post("/create", response_model=WorkflowResponseModel)
async def create_workflow(
    workflow: WorkflowCreateModel,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Crée un nouveau workflow"""
    # Créer le workflow
    workflow_data = workflow.dict()
    db_workflow = WorkflowService.create_workflow(db, workflow_data, current_user)
    
    if not db_workflow:
        # Récupérer l'abonnement pour vérifier les quotas
        subscription = None
        if current_user.company_id:
            subscription = SubscriptionService.get_company_subscription(db, current_user.company_id)
        else:
            subscription = SubscriptionService.get_user_subscription(db, current_user.id)
        
        if not subscription or not subscription.is_active:
            raise HTTPException(
                status_code=400,
                detail="Vous devez avoir un abonnement actif pour créer des workflows"
            )
        
        # Si l'utilisateur est solo, vérifier le quota de workflows
        if current_user.is_solo and subscription.max_workflows is not None:
            # Compter les workflows existants
            workflow_count = db.query(Workflow).filter(Workflow.owner_id == current_user.id).count()
            
            if workflow_count >= subscription.max_workflows:
                raise HTTPException(
                    status_code=400,
                    detail=f"Vous avez atteint la limite de {subscription.max_workflows} workflows pour votre abonnement"
                )
        
        raise HTTPException(
            status_code=500,
            detail="Erreur lors de la création du workflow"
        )
    
    return WorkflowService.workflow_to_dict(db_workflow)

@router.get("/{user_id}", response_model=List[WorkflowResponseModel])
async def get_user_workflows(
    user_id: str,
    current_user: User = Depends(get_current_user),
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """Récupère tous les workflows d'un utilisateur"""
    # Vérifier que l'utilisateur demande ses propres workflows ou est de la même entreprise
    if (current_user.id != user_id and
        (not current_user.company_id or current_user.company_id != db.query(User).filter(User.id == user_id).first().company_id)):
        raise HTTPException(
            status_code=403,
            detail="Vous n'êtes pas autorisé à consulter ces workflows"
        )
    
    workflows = WorkflowService.get_user_workflows(db, user_id, skip, limit)
    
    return [WorkflowService.workflow_to_dict(workflow) for workflow in workflows]

@router.get("/company/{company_id}", response_model=List[WorkflowResponseModel])
async def get_company_workflows(
    company_id: str,
    current_user: User = Depends(get_current_user),
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """Récupère tous les workflows partagés d'une entreprise"""
    # Vérifier que l'utilisateur appartient à l'entreprise
    if current_user.company_id != company_id:
        raise HTTPException(
            status_code=403,
            detail="Vous n'êtes pas autorisé à consulter ces workflows"
        )
    
    # Récupérer les workflows de l'entreprise qui sont partagés ou dont l'utilisateur est propriétaire
    workflows = db.query(Workflow).filter(
        (Workflow.company_id == company_id) & 
        ((Workflow.is_shared == True) | (Workflow.owner_id == current_user.id))
    ).offset(skip).limit(limit).all()
    
    return [WorkflowService.workflow_to_dict(workflow) for workflow in workflows]

@router.get("/detail/{workflow_id}", response_model=WorkflowResponseModel)
async def get_workflow_detail(
    workflow_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Récupère les détails d'un workflow spécifique"""
    # Récupérer le workflow
    workflow = WorkflowService.get_workflow(db, workflow_id)
    if not workflow:
        raise HTTPException(status_code=404, detail="Workflow non trouvé")
    
    # Vérifier l'accès
    if not WorkflowService.check_user_access(db, workflow_id, current_user.id):
        raise HTTPException(
            status_code=403,
            detail="Vous n'êtes pas autorisé à consulter ce workflow"
        )
    
    return WorkflowService.workflow_to_dict(workflow)

@router.put("/update/{workflow_id}", response_model=WorkflowResponseModel)
async def update_workflow(
    workflow_id: str,
    workflow: WorkflowUpdateModel,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Met à jour un workflow existant"""
    # Vérifier que l'utilisateur a accès au workflow
    db_workflow = WorkflowService.get_workflow(db, workflow_id)
    if not db_workflow:
        raise HTTPException(status_code=404, detail="Workflow non trouvé")
    
    if not WorkflowService.check_user_access(db, workflow_id, current_user.id):
        raise HTTPException(
            status_code=403,
            detail="Vous n'êtes pas autorisé à modifier ce workflow"
        )
    
    # Seul le propriétaire peut modifier certains attributs
    if current_user.id != db_workflow.owner_id:
        for attr in ['name', 'is_template']:
            if getattr(workflow, attr) is not None:
                raise HTTPException(
                    status_code=403,
                    detail=f"Seul le propriétaire peut modifier l'attribut '{attr}'"
                )
    
    # Mettre à jour le workflow
    workflow_data = {k: v for k, v in workflow.dict().items() if v is not None}
    updated_workflow = WorkflowService.update_workflow(db, workflow_id, workflow_data)
    
    if not updated_workflow:
        raise HTTPException(
            status_code=500,
            detail="Erreur lors de la mise à jour du workflow"
        )
    
    return WorkflowService.workflow_to_dict(updated_workflow)

@router.delete("/delete/{workflow_id}")
async def delete_workflow(
    workflow_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Supprime un workflow"""
    # Vérifier que l'utilisateur est le propriétaire du workflow
    db_workflow = WorkflowService.get_workflow(db, workflow_id)
    if not db_workflow:
        raise HTTPException(status_code=404, detail="Workflow non trouvé")
    
    if db_workflow.owner_id != current_user.id:
        raise HTTPException(
            status_code=403,
            detail="Seul le propriétaire peut supprimer un workflow"
        )
    
    # Supprimer le workflow
    result = WorkflowService.delete_workflow(db, workflow_id)
    
    if not result:
        raise HTTPException(
            status_code=500,
            detail="Erreur lors de la suppression du workflow"
        )
    
    return {"status": "success", "message": "Workflow supprimé avec succès"}

File: hooks/api/workflows.js
/**
 * Service API pour la gestion des workflows
 * Correspond aux routes /workflows dans le backend
 */
const workflowsService = (client) => {
  return {
    /**
     * Créer un nouveau workflow
     * @param {Object} workflowData - Données du workflow
     * @param {string} workflowData.name - Nom du workflow
     * @param {string} workflowData.description - Description (optionnel)
     * @param {Array} workflowData.nodes - Noeuds du workflow (optionnel)
     * @param {Array} workflowData.edges - Connexions du workflow (optionnel)
     * @param {boolean} workflowData.isShared - Si le workflow est partagé (optionnel)
     * @param {boolean} workflowData.isTemplate - Si le workflow est un modèle (optionnel)
     * @returns {Promise<Object>} - Le workflow créé
     */
    createWorkflow: (workflowData) => {
      const apiData = {
        name: workflowData.name,
        description: workflowData.description,
        nodes: workflowData.nodes || [],
        edges: workflowData.edges || [],
        is_shared: workflowData.isShared || false,
        is_template: workflowData.isTemplate || false
      };
      
      if (workflowData.createdAt) {
        apiData.createdAt = workflowData.createdAt;
      }
      
      return client.post('/workflows/create', apiData);
    },
    
    /**
     * Récupérer les workflows d'un utilisateur
     * @param {string} userId - ID de l'utilisateur
     * @param {Object} options - Options de pagination
     * @param {number} options.skip - Nombre d'éléments à ignorer
     * @param {number} options.limit - Nombre maximum d'éléments à récupérer
     * @returns {Promise<Array>} - Liste des workflows de l'utilisateur
     */
    getUserWorkflows: (userId, options = {}) => {
      const skip = options.skip || 0;
      const limit = options.limit || 100;
      
      return client.get(`/workflows/${userId}`, {
        params: { skip, limit }
      });
    },
    
    /**
     * Récupérer les workflows partagés d'une entreprise
     * @param {string} companyId - ID de l'entreprise
     * @param {Object} options - Options de pagination
     * @param {number} options.skip - Nombre d'éléments à ignorer
     * @param {number} options.limit - Nombre maximum d'éléments à récupérer
     * @returns {Promise<Array>} - Liste des workflows partagés de l'entreprise
     */
    getCompanyWorkflows: (companyId, options = {}) => {
      const skip = options.skip || 0;
      const limit = options.limit || 100;
      
      return client.get(`/workflows/company/${companyId}`, {
        params: { skip, limit }
      });
    },
    
    /**
     * Récupérer les détails d'un workflow
     * @param {string} workflowId - ID du workflow
     * @returns {Promise<Object>} - Détails du workflow
     */
    getWorkflowDetail: (workflowId) => 
      client.get(`/workflows/detail/${workflowId}`),
    
    /**
     * Mettre à jour un workflow
     * @param {string} workflowId - ID du workflow
     * @param {Object} workflowData - Données à mettre à jour
     * @param {string} workflowData.name - Nom (optionnel)
     * @param {string} workflowData.description - Description (optionnel)
     * @param {Array} workflowData.nodes - Noeuds (optionnel)
     * @param {Array} workflowData.edges - Connexions (optionnel)
     * @param {boolean} workflowData.isShared - Si le workflow est partagé (optionnel)
     * @param {boolean} workflowData.isTemplate - Si le workflow est un modèle (optionnel)
     * @returns {Promise<Object>} - Le workflow mis à jour
     */
    updateWorkflow: (workflowId, workflowData) => {
      const apiData = {};
      
      if (workflowData.name !== undefined) apiData.name = workflowData.name;
      if (workflowData.description !== undefined) apiData.description = workflowData.description;
      if (workflowData.nodes !== undefined) apiData.nodes = workflowData.nodes;
      if (workflowData.edges !== undefined) apiData.edges = workflowData.edges;
      if (workflowData.isShared !== undefined) apiData.is_shared = workflowData.isShared;
      if (workflowData.isTemplate !== undefined) apiData.is_template = workflowData.isTemplate;
      
      return client.put(`/workflows/update/${workflowId}`, apiData);
    },
    
    /**
     * Supprimer un workflow
     * @param {string} workflowId - ID du workflow à supprimer
     * @returns {Promise<Object>} - Confirmation de suppression
     */
    deleteWorkflow: (workflowId) => 
      client.delete(`/workflows/delete/${workflowId}`),
      
    /**
     * Dupliquer un workflow existant
     * @param {string} workflowId - ID du workflow à dupliquer
     * @param {Object} options - Options de duplication
     * @param {string} options.newName - Nouveau nom (optionnel)
     * @returns {Promise<Object>} - Le nouveau workflow créé
     */
    duplicateWorkflow: (workflowId, options = {}) => 
      client.post(`/workflows/duplicate/${workflowId}`, options),
      
    /**
     * Exporter un workflow au format JSON
     * @param {string} workflowId - ID du workflow à exporter
     * @returns {Promise<Object>} - Données du workflow au format JSON
     */
    exportWorkflow: (workflowId) => 
      client.get(`/workflows/export/${workflowId}`),
      
    /**
     * Importer un workflow depuis un fichier JSON
     * @param {Object} workflowData - Données du workflow à importer
     * @returns {Promise<Object>} - Le workflow importé
     */
    importWorkflow: (workflowData) => 
      client.post('/workflows/import', workflowData)
  };
};

export default workflowsService;